### PR TITLE
Settings Save button disable/enable

### PR DESCRIPTION
**Pull request by psealock**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes Partial fix for https://github.com/woocommerce/woocommerce/issues/47393

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to the all Settings screens and see the Save button is disabled. Make any change to the form and see that the Save button becomes enabled.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Update Settings to disable Save button unless modifications are made.

#### Comment 

</details>

### CODE REVIEW

**Review by moon0326 — APPROVED** [review]
Nice approach! LGTM and tested well 👍 🚀

**Review by chihsuan — COMMENTED** [review]
Thanks for working on this and adding the tests! 👍 

I confirmed that the Save button is disabled when no modifications are made and enabled when modifications are made.

However, if we reset the form after making modifications, the Save button is still enabled. What do you think disabling the Save button again when the form is reset?

It seems that we can achieve this by storing the initial values of the form fields and comparing them with the current values whenever an input event occurs.


```js
$(document).ready(function() {
    // Store the initial values of the form fields
    var initialFormState = $('#exampleForm').serialize();

    // Function to check if the form values have changed
    function checkFormChanges() {
        var currentFormState = $('#exampleForm').serialize();
        if (currentFormState === initialFormState) {
            $('#saveButton').prop('disabled', true);
        } else {
            $('#saveButton').prop('disabled', false);
        }
    }

    // Listen for changes in any of the form fields
    $('#exampleForm input').on('input', checkFormChanges);
});
```

**Comment by adrianduffell** [comment]
> What do you think disabling the Save button again when the form is reset?

That's a super idea @chihsuan! However I suggest we hold on it until we get feedback about the PR from developers (there will be communications about the change to mitigate risks to 3PD customization). I like that we're using an existing implementation to detect changes so I'd lean towards keeping our surface area small at first 🤔

**Comment by chihsuan** [comment]
https://github.com/woocommerce/woocommerce/assets/4344253/61fcb36c-b07e-443f-a916-9f4f78a0c041

Hey @psealock, I tested again with the latest code. I noticed that when woocommerce payment is installed, the Save changes button is not disabled when the page loads. I suspect WCPay changes the child elements after the observer is attached. 🤔

**Comment by psealock** [comment]
Good spotting @chihsuan. Payments indeed adds elements to the form via slotFill:

image

The idea behind adding the observer was to detect when form elements are added, [so I updated the observer](https://github.com/woocommerce/woocommerce/pull/47444/commits/2ae4d9b1ab1c3753f885672678fb6600586b2acc) to only re-add events if form inputs were added.

I have to then automatically remove the disabled attribute on the Save button because sometimes those inputs aren't required. See the case of Subscriptions where a Plan can be added, but no input is required before being able to save

image

**Review by chihsuan — APPROVED** [review]
Thanks for addressing feedback. @psealock  I confirmed Payment settings are working as expected now. 👍 

I tested all tabs, and it looks like **Settings > Shipping > Local Pickup:** Save changes button is not disabled on load as @ilyasfoo commented here  pdibGW-3lx-p2#comment-2539.

Other than that, everything else looks good to me. Pre-approving, thanks! 🙂

**Comment by psealock** [comment]
Thanks for the review @chihsuan, I handled the Local Pickup screen in [handle local pickup](https://github.com/woocommerce/woocommerce/pull/47444/commits/3ccf95a12053a5d443bf172f141fb6fe38f95adc)

**Review by chihsuan — APPROVED** [review]

File: plugins/woocommerce-blocks/tests/e2e/tests/local-pickup/local-pickup.merchant.block_theme.spec.ts [hunk 1]
```diff
@@ -220,7 +220,6 @@ test.describe( 'Merchant → Local Pickup Settings', () => {
 		await localPickupUtils.setLocalPickupTitle(
 			'Edited from settings page'
 		);
-		await localPickupUtils.saveLocalPickupSettings();
```
Comment (psealock) [inline plugins/woocommerce-blocks/tests/e2e/tests/local-pickup/local-pickup.merchant.block_theme.spec.ts:223]: The line above is already responsible for saving, so good to remove here.

**Pull request merged**